chpl/chpl-service/src/main/java/gov/healthit/chpl/manager/rules/developer/DeveloperEditTransparencyAttestationValidation.java

```diff
@@ -0,0 +1,63 @@
+package gov.healthit.chpl.manager.rules.developer;
+
+import java.util.Objects;
+
+import org.ff4j.FF4j;
+
+import gov.healthit.chpl.FeatureList;
+import gov.healthit.chpl.dto.DeveloperACBMapDTO;
+import gov.healthit.chpl.dto.DeveloperDTO;
+import gov.healthit.chpl.manager.rules.ValidationRule;
+import gov.healthit.chpl.permissions.ResourcePermissions;
+import gov.healthit.chpl.util.ErrorMessageUtil;
+
+public class DeveloperEditTransparencyAttestationValidation extends ValidationRule<DeveloperValidationContext> {
+    private FF4j ff4j;
+    private ResourcePermissions resourcePermissions;
+
+    public DeveloperEditTransparencyAttestationValidation(final FF4j ff4j,
+            final ResourcePermissions resourcePermissions) {
+        this.ff4j = ff4j;
+        this.resourcePermissions = resourcePermissions;
+    }
+
+    /**
+     * ROLE_ACB cannot edit Transparency Attestation as per OCD-3164
+     */
+    @Override
+    public boolean isValid(DeveloperValidationContext context) {
+        ErrorMessageUtil msgUtil = context.getErrorMessageUtil();
+        DeveloperDTO updatedDev = context.getDeveloperDTO();
+        if (ff4j.check(FeatureList.EFFECTIVE_RULE_DATE_PLUS_ONE_WEEK)) {
+            if (resourcePermissions.isUserRoleAcbAdmin()) {
+                if (isTransparencyAttestationUpdated(context.getBeforeDev(), updatedDev)) {
+                    getMessages().add(msgUtil.getMessage("developer.transparencyAttestationEditNotAllowedForRoleACB"));
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private static boolean isTransparencyAttestationUpdated(final DeveloperDTO original, final DeveloperDTO changed) {
+        if ((original.getTransparencyAttestationMappings() != null
```

Comment:
What if we changed "getTransparencyAttestationMappings" to return a list that was empty if there weren't any, instead of null? Then this entire set of things could be reduced to `if (original.getmaps.size != changed.getmaps.size)` and we don't have to worry about all the null checks?

Reply:
Seems like a good idea. I really like appreciate these kinds of comments the most, but:
-This isn't my code nor is it directly related to the ticket, thus, my focus was on refactoring the code into using the validation factory paradigm, and at the same time, on not changing the actual implementation. This way, code we know works, should still work in regression testing, especially since these changes are out of scope.
-What if something else calls getTransparencyAttestationMappings and depends on how it is now? What about implied verbal contracts of getters in general?
--To expand on my own questions: For the first one it's probably not the case that something depends on null, and if it does, maybe that's a problem in some circles. And the seconds one, that's a philosphy type thing too.

That being said, as usual, I'm not against it. We are going to regression test anyway. I will just open the discussion up to the second reviewer before I implement.

Reply:
Note: This code length could probably also be reduced using Objects.equals() and that would add the case of double null as well

Reply:
I'm fine not changing this for this PR, but I don't know if we'll have any other ticket that'd cover this section of the code any time soon. Your ticket is related to Transparency Attestations, so it's at least tangentially appropriate here.

Part of our process is to improve things as we come across them; maybe see how many references there are to this method in the code base, and if it's, like 1 or 2 that'd be a reasonable thing to update, but if it's 5+ that might make a difference

Reply:
If it's quick and the test cases will already regression test whatever is affected then I'm all for changing it. Agree if there's a lot of references or if it turns into one thing after another that gets changed then I wouldn't bother.

Reply:
Good points - sure thing, thanks

Reply:
There are 15 other independent calls to getTransparencyAttestationMappings() in the codebase so now I'm a bit hesitant to override it? As per your comment @andlar "maybe see how many references there are to this method in the code base, and if it's, like 1 or 2 that'd be a reasonable thing to update, but if it's 5+ that might make a difference" maybe we shouldn't change its impl then.
I looked into Objects.equals for the null parts and it seemed like it could work on this object (since it's a list) if we also overrode the equals method for the type (DeveloperACBMapDTO) to basically just chek size. Then we could use that as a one liner for the first set of checks. But, that's a no go because it also returns true if both are null which is great for property checks but would cause our loop to run on double null unless we add a condition to the else (which would we could to solve it). But, i'm not sure this is any more clear then the current impl. It's kind of convoluted at that point. 
I also looked into simplifyihng the else condition into one line of CollectionUtils.isEqualCollection(original, changed). The only concern there is that for now we are only checking getAcbName and getTransparencyAttestation where as isEqualCollection would check every property, and that probably will not get the result we like (and may be overkill). So would need testing. So anyway, I know I just said a lot but I'm not so sure any changes are neccessary or better at this point. But here's the options (besides overriding the getter) What do you guys think? I can look into/test if you think it's worth the time.

```
    private static boolean isTAUpdatedAlt(final DeveloperDTO original, final DeveloperDTO changed) {
        // override equals to just check size
        if (!Objects.equals(original.getTransparencyAttestationMappings(),
                changed.getTransparencyAttestationMappings())) {
            return true;
        }
        if (original.getTransparencyAttestationMappings() != null
                && changed.getTransparencyAttestationMappings() != null) {
            for (DeveloperACBMapDTO originalMapping : original.getTransparencyAttestationMappings()) {
                for (DeveloperACBMapDTO changedMapping : changed.getTransparencyAttestationMappings()) {
                    if (!Objects.equals(originalMapping.getAcbName(), changedMapping.getAcbName())
                            || !Objects.equals(originalMapping.getTransparencyAttestation(),
                                    changedMapping.getTransparencyAttestation())) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private static boolean isTAUpdatedAlt2(final DeveloperDTO original, final DeveloperDTO changed) {
        if ((original.getTransparencyAttestationMappings() != null
                && changed.getTransparencyAttestationMappings() == null)
                || (original.getTransparencyAttestationMappings() == null
                        && changed.getTransparencyAttestationMappings() != null)
                || (original.getTransparencyAttestationMappings().size() != changed.getTransparencyAttestationMappings()
                        .size())) {
            return true;
        }
        if (original.getTransparencyAttestationMappings() != null
                && changed.getTransparencyAttestationMappings() != null) {
            return CollectionUtils.isEqualCollection(original.getTransparencyAttestationMappings(),
                    changed.getTransparencyAttestationMappings());
        }
        return false;

    }

    private static boolean isTAUpdatedAltMixed(final DeveloperDTO original, final DeveloperDTO changed) {
        // override equals to just check size
        if (!Objects.equals(original.getTransparencyAttestationMappings(),
                changed.getTransparencyAttestationMappings())) {
            return true;
        }
        if (original.getTransparencyAttestationMappings() != null
                && changed.getTransparencyAttestationMappings() != null) {
            return CollectionUtils.isEqualCollection(original.getTransparencyAttestationMappings(),
                    changed.getTransparencyAttestationMappings());
        }
        return false;
    }
```

Reply:
Looks like the best path forward is to do nothing with this set of things just yet. That's too many other places, and I don't see a lot of need in overriding all these other things either. I'm going to resolve this conversation and approve the PR as it stands so we can think about getting this part into the testing state

Reply:
Here's what the getter version would look like (I did this in like 5 secs so might be a type):
    private static boolean isTAUpdatedGetterAlt(final DeveloperDTO original, final DeveloperDTO changed) {

>         // override getTransparencyAttestationMappings to return an empty list
>         // vs null
>         if (original.getTransparencyAttestationMappings().size() != changed.getTransparencyAttestationMappings()
>                 .size()) {
>             return true;
>         }
>         if (original.getTransparencyAttestationMappings() != null
>                 && changed.getTransparencyAttestationMappings() != null) {
>             for (DeveloperACBMapDTO originalMapping : original.getTransparencyAttestationMappings()) {
>                 for (DeveloperACBMapDTO changedMapping : changed.getTransparencyAttestationMappings()) {
>                     if (!Objects.equals(originalMapping.getAcbName(), changedMapping.getAcbName())
>                             || !Objects.equals(originalMapping.getTransparencyAttestation(),
>                                     changedMapping.getTransparencyAttestation())) {
>                         return true;
>                     }
>                 }
>             }
>         }
>         return false;
>     }

Reply:
Sure thing - good liittle thought excercise if nothing else